test/es-module/test-esm-dynamic-import-assertion.mjs

```diff
@@ -0,0 +1,41 @@
+// Flags: --experimental-json-modules
+import '../common/index.mjs';
+import { rejects, strictEqual } from 'assert';
+
+const jsModuleDataUrl = 'data:text/javascript,export{}';
+
+await rejects(
+  import(`data:text/javascript,import${JSON.stringify(jsModuleDataUrl)}assert{type:"json"}`),
+  { code: 'ERR_FAILED_IMPORT_ASSERTION' }
+);
+
+await rejects(
+  import(jsModuleDataUrl, { assert: { type: 'json' } }),
+  { code: 'ERR_FAILED_IMPORT_ASSERTION' }
+);
+
+{
+  const [secret0, secret1] = await Promise.all([
+    import('../fixtures/experimental.json'),
+    import(
+      '../fixtures/experimental.json',
+      { assert: { type: 'json' } }
+    ),
+  ]);
+
+  strictEqual(secret0.default.ofLife, 42);
+  strictEqual(secret1.default.ofLife, 42);
```

Comment:
@bmeck is this what you had in mind when asking for a test with a race condition? I suppose it's not, I didn't change the cache implementation yet this test passes.

Reply:
you should test references not primitives, see if the module namespace objects are different or if the `secret0.default === secret1.default` since they have reference semantics.

Reply:
the purpose of the test should be to see if the module was loaded more than once. checking references, checking mutation of one and seeing if the other is effected, etc.

Reply:
I've added the tests, and they pass.

Reply:
I think that with the current code, import assertions have no effect on the cache.

Reply:
Another race condition that we can check is:

```js
Promise.all([
  import('./not-a-json.js', { assert: { type: 'json' } }),
  import('./not-a-json.js')
])
```

Reply:
> I think that with the current code, import assertions have no effect on the cache.

I'm looking for a test case that would have an effect on the ESM cache so I can update the implementation accordingly.

Reply:
It depends on what is the behavior that we want, and I think there are different opinions on that.

Reply:
in general due to needing to force a race you would want to use a loader and manually resolve the winner of the race only after both module jobs have started.

Reply:
With the new ESM loader implementation that just landed, the above test indeed created a race condition so I was able to move this forward :)